### PR TITLE
feat(2d): add mask

### PR DESCRIPTION
**Pull request by singlecoder**
### Please check if the PR fulfills these requirements

- [ ] The commit message follows our [guidelines](https://github.com/oasis-engine/engine/blob/main/.github/COMMIT_MESSAGE_CONVENTION.md)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

### CODE REVIEW

File: packages/core/src/2d/enums/SpriteMaskLayer.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
+/**
+ * Sprite mask layer.
+ */
+export enum SpriteMaskLayer {
+  Layer0 = 0x1,
```
Comment (GuoLei1990) [inline packages/core/src/2d/enums/SpriteMaskLayer.ts:5]: write simple comments like "mask layer 0."

File: packages/core/src/2d/sprite/SpriteRenderer.ts [hunk 2]
```diff
@@ -93,6 +99,28 @@ export class SpriteRenderer extends Renderer {
...
+  }
+
+  set maskInteraction(value: SpriteMaskInteraction) {
+
+  }
+
+  /**
+   * The mask layer the sprite renderer belongs to.
+   */
+  get maskLayer(): number {
```
Comment (GuoLei1990) [inline packages/core/src/2d/sprite/SpriteRenderer.ts:116]: return type is enum

File: packages/core/src/2d/enums/SpriteMaskLayer.ts (hunk 1 above)
Comment (singlecoder) [reply packages/core/src/2d/enums/SpriteMaskLayer.ts:6]: > write simple comments like "mask layer 0."

done

File: packages/core/src/2d/sprite/SpriteMask.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+   * The mask layers the sprite mask influence to.
+   */
+  get influenceLayers(): number {
+    return this._influenceLayers;
+  }
+
+  set influenceLayers(value: number) {}
+
+  /**
+   * Create a sprite mask instance.
```
Comment (GuoLei1990) [inline packages/core/src/2d/sprite/SpriteMask.ts:46]: @internal

File: packages/core/src/2d/sprite/SpriteRenderer.ts (hunk 2 above)
Comment (singlecoder) [reply packages/core/src/2d/sprite/SpriteRenderer.ts:116]: > return type is enum

Number is better，SpriteMaskLayer.Layer0 | SpriteMaskLayer.Layer1

File: packages/core/src/Layer.ts [hunk 4]
```diff
@@ -2,38 +2,72 @@
  * Layer, used for bit operations.
  */
 export enum Layer {
+  /** layer 0. */
```
Comment (GuoLei1990) [inline packages/core/src/Layer.ts:5]: ```/** Layer 0. */```  not  ```/** layer 0. */```

File: packages/core/src/2d/sprite/SpriteMask.ts [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+import { Sprite } from "./Sprite";
+import "./SpriteMaskMaterial";
+
+/**
+ * A component for masking Sprites.
+ */
+export class SpriteMask extends Component {
+  private static _tempVec3: Vector3 = new Vector3();
+  static textureProperty: ShaderProperty = Shader.getPropertyByName("u_maskTexture");
+  static alphaCutoffProperty: ShaderProperty = Shader.getPropertyByName("u_alphaCutoff");
```
Comment (GuoLei1990) [inline packages/core/src/2d/sprite/SpriteMask.ts:21]:
```typescript  
/** @internal */
  static _textureProperty: ShaderProperty = Shader.getPropertyByName("u_maskTexture");
  /** @internal */
  static _alphaCutoffProperty: ShaderProperty = Shader.getPropertyByName("u_alphaCutoff");
```

and should before private

File: packages/core/src/2d/sprite/SpriteMask.ts [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+  }
+
+  set influenceLayers(value: number) {
+    this._influenceLayers = value;
+  }
+
+  /**
+   * Get the material.
+   */
+  get material(): Material {
```
Comment (GuoLei1990) [inline packages/core/src/2d/sprite/SpriteMask.ts:77]: delete this and extend `SpriteMask` form `Renderer`

File: packages/core/src/RenderPipeline/SpriteMaskManager.ts [hunk 7]
```diff
@@ -0,0 +1,403 @@
...
+  MeshTopology,
+  SubMesh,
+  VertexElement,
+  VertexElementFormat
+} from "../graphic";
+import { BufferMesh } from "../mesh";
+import { Shader, StencilOperation } from "../shader";
+import { SystemInfo } from "../SystemInfo";
+import { ClassPool } from "./ClassPool";
+import { SpriteMaskElement } from "./SpriteMaskElement";
```
Comment (GuoLei1990) [inline packages/core/src/RenderPipeline/SpriteMaskManager.ts:19]: clear your imports, not import * by "index.ts"

File: packages/core/src/2d/sprite/SpriteMask.ts [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+    this._isWorldMatrixDirty = entity.transform.registerWorldChangeFlag();
+    this._material = this._createMaterial();
+  }
+
+  _onEnable(): void {
+    const manager = SpriteMaskManager.getInstance(this.engine);
+    manager.addMask(this);
+  }
+
+  _onDisable(): void {
```
Comment (GuoLei1990) [inline packages/core/src/2d/sprite/SpriteMask.ts:95]: add @internal

File: packages/core/src/2d/sprite/SpriteMask.ts [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+    const manager = SpriteMaskManager.getInstance(this.engine);
+    manager.removeMask(this);
+  }
+
+  getElement(): SpriteMaskElement {
```
Comment (GuoLei1990) [inline packages/core/src/2d/sprite/SpriteMask.ts:100]: add @internal and named "_getElement"

File: packages/core/src/2d/sprite/SpriteMask.ts [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+      }
+
+      this._isSpriteDirty = false;
+      this._isWorldMatrixDirty.flag = false;
+    }
+
+    const material = this._material;
+    const shaderData = material.shaderData;
+    shaderData.setTexture(SpriteMask.textureProperty, texture);
+    shaderData.setFloat(SpriteMask.alphaCutoffProperty, this.alphaCutoff);
```
Comment (GuoLei1990) [inline packages/core/src/2d/sprite/SpriteMask.ts:134]: two properties is not belong material,they belong SpriteMask

File: packages/core/src/Layer.ts (hunk 4 above)
Comment (singlecoder) [reply packages/core/src/Layer.ts:5]: done

File: packages/core/src/RenderPipeline/SpriteMaskManager.ts [hunk 11]
```diff
@@ -0,0 +1,406 @@
...
+    }
+
+    if (curLen === 0) {
+      for (let i = 0; i < newLen; ++i) {
+        this._preDrawMask(newElements[i], true);
+      }
+      return;
+    }
+
+    const repeatMasks = SpriteMaskManager._tempMasks;
```
Comment (GuoLei1990) [inline packages/core/src/RenderPipeline/SpriteMaskManager.ts:202]: Bad performance!!!

File: packages/core/src/2d/sprite/SpriteMask.ts [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+  /**
+   * The mask layers the sprite mask influence to.
+   */
+  get influenceLayers(): number {
+    return this._influenceLayers;
+  }
+
+  set influenceLayers(value: number) {
+    this._influenceLayers = value;
+  }
```
[the remark this answers is not in the document]
Comment (singlecoder) [reply packages/core/src/2d/sprite/SpriteMask.ts]: > If no logic use `public influenceLayers:number` directionly

done

File: packages/core/src/RenderPipeline/SpriteMaskManager.ts [hunk 13]
```diff
@@ -0,0 +1,406 @@
...
+import { SpriteMaskElement } from "./SpriteMaskElement";
+
+export class SpriteMaskManager {
+  private static _instance: SpriteMaskManager = null;
+  private static _tempMasks: Set<SpriteMask> = new Set<SpriteMask>();
+  /** The maximum number of vertex. */
+  private static MAX_VERTEX_COUNT: number = 4096;
+  private static _canUploadSameBuffer: boolean = !SystemInfo._isIos();
+
+  static getInstance(engine: Engine): SpriteMaskManager {
```
[the remark this answers is not in the document]
Comment (singlecoder) [reply packages/core/src/RenderPipeline/SpriteMaskManager.ts]: > Multi engine will cause bug.

done

File: packages/core/src/Engine.ts [hunk 14]
```diff
@@ -88,6 +100,13 @@ export class Engine extends EventDispatcher {
     return this._sceneManager;
   }
 
+  /**
+   * Get the sprite mask manager.
+   */
+  get spriteMaskManager(): SpriteMaskManager {
```
Comment (GuoLei1990) [inline packages/core/src/Engine.ts:106]: delete this, SpriteMaskManager is internal

**Review by GuoLei1990 — APPROVED** [review]
+1

**Pull request merged**